lib/compiler/src/beam_ssa_throw.erl

```diff
@@ -0,0 +1,495 @@
+si_is([#b_set{op=call,
+              dst=Dst,
+              args=[#b_remote{mod=#b_literal{val=erlang},
+                              name=#b_literal{val=throw},
+                              arity=1}, _Term]}],
+      Id, _Lbl, #b_ret{arg=Dst}, Lst, Gst) ->
+    %% Tail throw, handled by caller. We'll need to visit this function again.
+    #gst{throws=Throws0} = Gst,
+    Throws = cerl_sets:add_element(Id, Throws0),
+    {Lst, Gst#gst{throws=Throws}};
+si_is([#b_set{op=call,dst=Dst,args=[#b_local{}=Callee | _]}],
+      Id, _Lbl, #b_ret{arg=Dst}, Lst, Gst) ->
+    %% Tail call, inherit our caller's handler.
+    {Lst, inherit_tlh(Id, Callee, Gst)};
+si_is([#b_set{op=call,dst=Dst,args=[#b_local{}=Callee | _]},
+       #b_set{op={succeeded,body},args=[Dst]}],
+      Id, _Lbl, #b_br{fail=?EXCEPTION_BLOCK}, Lst, Gst) ->
+    %% No local handler, inherit our caller's top-level handlers.
+    {Lst, inherit_tlh(Id, Callee, Gst)};
+si_is([#b_set{op=call,dst=Dst,args=[#b_local{}=Callee | _]},
+       #b_set{op={succeeded,body},args=[Dst]}],
+      Id, _Lbl, #b_br{fail=Fail}, Lst, Gst) ->
+    %% Local handler, register it as a top-level handler with our callee.
+    HandlerId = {Id, Fail},
+    {Lst, add_tlh(HandlerId, Callee, Lst, Gst)};
+si_is([#b_set{} | Is], Id, Lbl, Last, Lst, Gst) ->
+    si_is(Is, Id, Lbl, Last, Lst, Gst);
+si_is([], _Id, _Lbl, _Last, Lst, Gst) ->
+    {Lst, Gst}.
+
+si_handler_start([#b_set{op=extract,
+                         dst=Dst,
+                         args=[_, #b_literal{val=Idx}]} | Is],
+                 Id, StartLbl, Last, Vars0, Lst, Gst) ->
+    none = element(1 + Idx, Vars0),             %Assertion.
+    Vars = setelement(1 + Idx, Vars0, Dst),
+    si_handler_start(Is, Id, StartLbl, Last, Vars, Lst, Gst);
+si_handler_start(Is, Id, StartLbl, Last, Vars, Lst0, Gst) ->
+    HandlerId = {Id, StartLbl},
+
+    #lst{blocks=Blocks,
+         predecessors=Preds,
+         suitability=Suitability,
+         handlers=Handlers0} = Lst0,
+
+    {_, _, Stacktrace} = Vars,
+
+    Handlers = case Suitability of
+                   #{ Stacktrace := {tentative, EndLbl} } ->
+                       %% This handler's suitability depends on what we throw,
+                       %% so we'll extract the blocks between handler start and
+                       %% the first use of the stack trace for later analysis.
+                       Path = beam_ssa:between(StartLbl, EndLbl, Preds, Blocks),
+                       Partition = si_copy_blocks(Path, Blocks, #{}),
+
+                       Handler = {tentative, StartLbl, Vars, Partition},
+                       Handlers0#{ HandlerId => Handler };
+                   #{} when Stacktrace =/= none ->
+                       %% Stack trace is used in an unsupported manner.
+                       Handlers0#{ HandlerId => unsuitable };
+                   #{} when Stacktrace =:= none ->
+                       Handlers0#{ HandlerId => suitable }
```

Comment:
Just for clarity, a comment that notes that the stacktrace is never used would be useful.